File: crates/step_sim/src/runner.rs
//! Simulation execution functionality
use super::agents::{AgentSet, MarketAgentSet};
use super::env::Env;
use super::market_env::MarketEnv;
use kdam::tqdm;
use rand_xoshiro::rand_core::SeedableRng;
use rand_xoshiro::Xoroshiro128StarStar;

/// Run a simulation for a fixed number of steps
///
/// Each step updates the state of the agents (who
/// in turn can submit instructions to the environment
/// and then update the environment state)
///
/// # Examples
///
/// ```
/// use bourse_de::{Env, sim_runner};
/// use bourse_de::agents::AgentSet;
/// use rand::RngCore;
///
/// // Dummy agent-type
/// struct Agents{}
///
/// impl AgentSet for Agents {
///     fn update<R: RngCore>(
///         &mut self, env: &mut Env, _rng: &mut R
///     ) {}
/// }
///
/// let mut env = bourse_de::Env::new(0, 1, 1_000, true);
/// let mut agents = Agents{};
///
/// // Run for 100 steps from seed 101
/// sim_runner(&mut env, &mut agents, 101, 100, true)
/// ```
///
/// # Arguments
///
/// - `env` - Simulation environment
/// - `agents` - Agent(s) implementing the [AgentSet] trait
/// - `seed` - Random seed
/// - `n_steps` - Number of simulation steps
/// - `show_progress` - Show progress bar
///
pub fn sim_runner<A: AgentSet>(
    env: &mut Env,
    agents: &mut A,
    seed: u64,
    n_steps: u64,
    show_progress: bool,
) {
    let mut rng = Xoroshiro128StarStar::seed_from_u64(seed);

    match show_progress {
        true => {
            for _ in tqdm!(0..n_steps) {
                agents.update(env, &mut rng);
                env.step(&mut rng);
            }
        }
        false => {
            for _ in 0..n_steps {
                agents.update(env, &mut rng);
                env.step(&mut rng);
            }
        }
    }
}

/// Run a simulation for a fixed number of steps
///
/// Each step updates the state of the agents (who
/// in turn can submit instructions to the environment
/// and then update the environment state)
///
/// # Examples
///
/// ```
/// use bourse_de::{MarketEnv, market_sim_runner};
/// use bourse_de::agents::MarketAgentSet;
/// use rand::RngCore;
///
/// // Dummy agent-type
/// struct Agents{}
///
/// impl MarketAgentSet for Agents {
///     fn update<R: RngCore, const M: usize, const N: usize>(
///         &mut self, _env: &mut MarketEnv<M, N>, _rng: &mut R
///     ) {}
/// }
///
/// let mut env = bourse_de::MarketEnv::<2>::new(0, [1, 1], 1_000, true);
/// let mut agents = Agents{};
///
/// // Run for 100 steps from seed 101
/// market_sim_runner(&mut env, &mut agents, 101, 100, true)
/// ```
///
/// # Arguments
///
/// - `env` - Simulation environment
/// - `agents` - Agent(s) implementing the [MarketAgentSet] trait
/// - `seed` - Random seed
/// - `n_steps` - Number of simulation steps
/// - `show_progress` - Show progress bar
///
pub fn market_sim_runner<A: MarketAgentSet, const M: usize, const N: usize>(
    env: &mut MarketEnv<M, N>,
    agents: &mut A,
    seed: u64,
    n_steps: u64,
    show_progress: bool,
) {
    let mut rng = Xoroshiro128StarStar::seed_from_u64(seed);

    match show_progress {
        true => {
            for _ in tqdm!(0..n_steps) {
                agents.update(env, &mut rng);
                env.step(&mut rng);
            }
        }
        false => {
            for _ in 0..n_steps {
                agents.update(env, &mut rng);
                env.step(&mut rng);
            }
        }
    }
}

File: src/bourse/step_sim/runner.py
"""
Discrete event simulation runner
"""
import typing

import numpy as np
import tqdm

import bourse


def run(
    env: typing.Union[bourse.core.StepEnv, bourse.core.StepEnvNumpy],
    agents: typing.Iterable,
    n_steps: int,
    seed: int,
    show_progress: bool = True,
    use_numpy: bool = False,
) -> typing.Dict[str, np.ndarray]:
    """
    Run a discrete event simulation for fixed number of steps

    Runs a discrete event simulation. Each step of the
    simulation agents submit transactions to the
    simulation environment. At the end of the transactions
    are randomly shuffled and process, updating the state of
    the market.

    Examples
    --------

    .. testsetup:: runner_docstring

       import bourse

       agents = []
       env = env = bourse.core.StepEnv(0, 0, 2, 1000)

    .. testcode:: runner_docstring

       market_data = bourse.step_sim.run(
           env,     # Simulation environment
           agents,  # List of agents
           50,      # Number of steps
           101      # Random seed
       )

    Parameters
    ----------
    env: bourse.core.StepEnv
        Step updating simulation environment
    agents: list or tuple
        Iterable containing initialised agents. Agents
        should have an ``update`` method that interacts
        with the simulation environment, or if ``use_numpy``
        is ``True`` then agents should return a tuple
        of Numpy array instructions. See
        :py:class:`bourse.step_sim.agents.base_agent.BaseAgent`
        and
        :py:class:`bourse.step_sim.agents.base_agent.BaseNumpyAgent`
        for more details.
    n_steps: int
        Number of simulation steps to run.
    seed: int
        Random seed.
    show_progress: bool, optional
        If ``True`` a progress bar will be displayed,
        default ``True``
    use_numpy: bool, optional
        If ``True`` use numpy api to for market state and
        to submit market instructions. Default ``False``

    Returns
    -------
    dict
        Dictionary containing level 2 market data with keys:

        - ``bid_price``: Bid price at each step
        - ``ask_price``: Ask price at each step
        - ``bid_vol``: Total bid volume at each step
        - ``ask_vol``: Total ask volume at each step
        - ``trade_vol``: Trade volume each step
        - ``bid_vol_<N>``: Bid volume at top 10 levels at each step
        - ``ask_vol_<N>``: Ask volume at top 10 levels at each step
        - ``n_bid_<N>``: Number of bid orders at top 10 levels at each step
        - ``n_ask_<N>``: Number of ask orders at top 10 levels at each step
    """

    if use_numpy:
        assert isinstance(env, bourse.core.StepEnvNumpy)
        assert all(
            [isinstance(a, bourse.step_sim.agents.BaseNumpyAgent) for a in agents]
        ), "Agents should implement bourse.step_sim.agents.BaseNumpyAgent"
    else:
        assert isinstance(env, bourse.core.StepEnv)
        assert all(
            [isinstance(a, bourse.step_sim.agents.BaseAgent) for a in agents]
        ), "Agents should implement bourse.step_sim.agents.BaseAgent"

    rng = np.random.default_rng(seed)
    it = tqdm.trange(n_steps, disable=not show_progress)

    if use_numpy:
        for _ in it:

            level_2_data = env.level_2_data()

            for agent in agents:
                instructions = agent.update(rng, level_2_data)
                env.submit_instructions(instructions)

            env.step()
    else:
        for _ in it:
            for agent in agents:
                agent.update(rng, env)

            env.step()

    return env.get_market_data()
